crates/sui-core/src/execution_cache/writeback_cache.rs

```diff
@@ -0,0 +1,1267 @@
+        do_fallback_lookup(
+            event_digests,
+            |digest| {
+                if let Some(events) = self.dirty.transaction_events.get(digest) {
+                    CacheResult::Hit(Some(events.1.clone()))
+                } else {
+                    CacheResult::Miss
+                }
+            },
+            |digests| self.store.multi_get_events(digests),
+        )
+    }
+
+    fn get_sui_system_state_object_unsafe(&self) -> SuiResult<SuiSystemState> {
+        get_sui_system_state(self)
+    }
+
+    fn get_marker_value(
+        &self,
+        object_id: &ObjectID,
+        version: SequenceNumber,
+        epoch_id: EpochId,
+    ) -> SuiResult<Option<MarkerValue>> {
+        match self.get_marker_value_cache_only(object_id, version, epoch_id) {
+            CacheResult::Hit(marker) => Ok(Some(marker)),
+            CacheResult::NegativeHit => Ok(None),
+            CacheResult::Miss => self.store.get_marker_value(object_id, &version, epoch_id),
+        }
+    }
+
+    fn get_latest_marker(
+        &self,
+        object_id: &ObjectID,
+        epoch_id: EpochId,
+    ) -> SuiResult<Option<(SequenceNumber, MarkerValue)>> {
+        match self.get_latest_marker_value_cache_only(object_id, epoch_id) {
+            CacheResult::Hit((v, marker)) => Ok(Some((v, marker))),
+            CacheResult::NegativeHit => {
+                panic!("cannot have negative hit when getting latest marker")
+            }
+            CacheResult::Miss => self.store.get_latest_marker(object_id, epoch_id),
+        }
+    }
+
+    fn get_lock(&self, _obj_ref: ObjectRef, _epoch_id: EpochId) -> SuiLockResult {
+        todo!()
+    }
+
+    fn _get_latest_lock_for_object_id(&self, _object_id: ObjectID) -> SuiResult<ObjectRef> {
+        todo!()
+    }
+
+    fn check_owned_object_locks_exist(&self, _owned_object_refs: &[ObjectRef]) -> SuiResult {
+        todo!()
+    }
+}
+
+impl ExecutionCacheWrite for WritebackCache {
+    #[instrument(level = "trace", skip_all)]
+    fn acquire_transaction_locks<'a>(
+        &'a self,
+        _epoch_id: EpochId,
+        _owned_input_objects: &'a [ObjectRef],
+        _tx_digest: TransactionDigest,
+    ) -> BoxFuture<'a, SuiResult> {
+        todo!()
+    }
+
+    #[instrument(level = "debug", skip_all)]
+    fn write_transaction_outputs(
+        &self,
+        epoch_id: EpochId,
+        tx_outputs: Arc<TransactionOutputs>,
+    ) -> BoxFuture<'_, SuiResult> {
+        async move {
+            let TransactionOutputs {
+                transaction,
+                effects,
+                markers,
+                written,
+                deleted,
+                wrapped,
+                events,
+                ..
+            } = &*tx_outputs;
+
+            // Deletions and wraps must be written first. The reason is that one of the deletes
+            // may be a child object, and if we write the parent object first, a reader may or may
```

Comment:
I don't fully understand the potential issue here.
If a later transaction needs to read an object at version V, it must wait until it's notified that V is ready. This notification won't happen until everything below finishes. So as long as we have MVCC, the order below shouldn't matter?

Reply:
If a transaction is enqueued in TxMgr, then yes, we will not currently wake it until all writes are finished. However that is not the only case - a transaction could arrive while we are still in this function, it would check if its input objects are available, and then run immediately instead of going into the queue, potentially before the child tombstone is written.

Additionally, i would eventually like to move to using a notify_read_object pattern and remove transaction manager, which will expose this issue even more.